### PR TITLE
Changing Number.NumberBuffer to carry a `Span<byte>` rather than a `Span<char>`

### PR DESCRIPTION
**Pull request by tannergooding**
As per https://github.com/dotnet/coreclr/pull/20873#issuecomment-437090892, I split the PR into multiple.

This is the first PR and changes `Number.NumberBuffer` to carry a `Span<byte>`, rather than a `Span<char>`. This will allow easier sharing with the `Utf8Formatter` and `Utf8Parser`.
* The span is already only handling the ASCII characters `0-9` and the null terminator character, so there shouldn't be any issues.

### CODE REVIEW

**Comment by tannergooding** [comment]
Working on getting perf numbers

File: src/System.Private.CoreLib/shared/System/Number.Formatting.cs [hunk 1]
```diff
@@ -1094,18 +1094,30 @@ private static unsafe bool TryInt32ToHexStr(int value, char hexBase, int digits,
...
 
-            char* dst = number.GetDigitsPointer();
+            byte* dst = number.GetDigitsPointer();
             while (--i >= 0)
                 *dst++ = *p++;
-            *dst = '\0';
+            *dst = (byte)('\0');
+        }
+
+        internal static unsafe byte* UInt32ToDecChars(byte* bufferEnd, uint value, int digits)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Number.Formatting.cs]: Is the `char*` variant still used after this is introduced?

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Number.Formatting.cs]: Yes, in the formatter. I think we could probably update this to be shareable though.

**Comment by tannergooding** [comment]
Added two new commits:
* The first adds some `number.CheckConsistency` calls
* The second ensures that the consistency checks are being met (the UTF16Formatter wasn't properly tracking the digit count)

**Comment by tannergooding** [comment]
@stephentoub: Perf numbers look good.
* [i7-7700_Base.csv.txt](https://github.com/dotnet/coreclr/files/2564454/i7-7700_Base.csv.txt)
* [i7-7700_Diff.csv.txt](https://github.com/dotnet/coreclr/files/2564455/i7-7700_Diff.csv.txt)

We seem to be generally slightly faster then previously.
* For the integer tests, we always have the same number of iterations. Perf is generally within the std-dev, but looks overall like it might be slightly faster
* For the floating-point tests, we generally squeeze out some additional iterations and also look to be overall faster

**Comment by stephentoub** [comment]
Excellent. Thank you.

**Pull request merged**